### PR TITLE
feat: DAH-2010: Upgrade to Typescript 5 and fortify types

### PR DESCRIPTION
**Pull request by chadbrokaw**
[DAH-2010](https://sfgovdt.jira.com/browse/DAH-2010)

[DAH-2010]: https://sfgovdt.jira.com/browse/DAH-2010?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: Gemfile
Comment (jimlin-sfgov) [inline Gemfile:26]: question: Is this intentional? Seems unusual to fold a major version ruby gem change into a Node package update.

File: app/javascript/api/types/rails/listings/RailsUnit.d.ts
Comment (jimlin-sfgov) [inline app/javascript/api/types/rails/listings/RailsUnit.d.ts:24]: thought: this seems like a code smell indicating that we're going against the data immutability principle of React.

File: app/javascript/hooks/useScript.tsx
Comment (jimlin-sfgov) [inline app/javascript/hooks/useScript.tsx:39]: suggestion: This entire file is written in a style that predates Typescript, and is used only to facilitate our hacky machine translation process. We can just ignore the entire file by adding [// @ts-nocheck](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#-ts-nocheck-in-typescript-files) to the top of the file/

File: app/javascript/modules/listingDetails/ListingDetailsUnitAccordions.tsx
Comment (jimlin-sfgov) [inline app/javascript/modules/listingDetails/ListingDetailsUnitAccordions.tsx:59]: suggestion: We can use the [record utility](https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeys-type) to save a few lines and remove the need for the `PropertType` type, e.g.
```js
const sortUnits = (units: RailsUnit[]): Record<RailsUnit["Unit_Type"], UnitType> => {
```

File: app/javascript/modules/listings/DirectoryHelpers.tsx
Comment (jimlin-sfgov) [inline app/javascript/modules/listings/DirectoryHelpers.tsx:233]: question: I'm seeing `stackedDataFxn` appear a lot in a other places as an untyped argument. Is it worth the effort to define this type in a more accessible place so that we can use it in more places?

File: Gemfile
Comment (chadbrokaw) [reply Gemfile:26]: Thank you for calling that out! This snuck in because I was getting a bunch of bundler issues with 5.6.8, upgrading to 6.4.2 solved those (and brought us in line with Partners). I extracted this into its own PR (#2073) so the separation is clearer

File: app/javascript/util/routeUtil.ts
Comment (chadbrokaw) [reply app/javascript/util/routeUtil.ts:52]: I'm not entirely sure what you mean here? Do you mean changing line 52 to be:
```suggestion
  return getLocalizedPath(currentPath, LanguagePrefix.English, queryString)
```

File: app/javascript/api/types/rails/listings/RailsUnit.d.ts
Comment (chadbrokaw) [reply app/javascript/api/types/rails/listings/RailsUnit.d.ts:24]: Sigh, I agree... however I prefer this over `any`. With that said, this should be refactored in another ticket to remove the looseness of this type, unless you see an easy way to do it?

File: app/javascript/hooks/useScript.tsx
Comment (chadbrokaw) [reply app/javascript/hooks/useScript.tsx:39]: I like this idea but we would also have to attach an eslint ignore for the ts comment itself. What if we just applied eslint rules to the file as needed? So the top of the file would have:
```
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-argument */
```

As an aside, this file could use a thorough refactor to be better in line with Typescript principles.

File: app/javascript/modules/listingDetails/ListingDetailsUnitAccordions.tsx
Comment (chadbrokaw) [reply app/javascript/modules/listingDetails/ListingDetailsUnitAccordions.tsx:59]: 😍 Amazing!

File: app/javascript/__tests__/modules/listingDetails/ListingDetailsUnitAccordions.test.tsx
Comment (chadbrokaw) [reply app/javascript/__tests__/modules/listingDetails/ListingDetailsUnitAccordions.test.tsx:35]: I'm not 100% sure, but I think what is happening is that the useReducer is not provided the ContextProps type, so the type checker is being more lenient with the starting state. 

I will look into this more.

File: app/javascript/api/types/rails/listings/RailsUnit.d.ts
Comment (jimlin-sfgov) [reply app/javascript/api/types/rails/listings/RailsUnit.d.ts:24]: I don't see an easy way, otherwise I would have offered a suggestion. I can create a ticket to look into the complexities of certain functions like `matchSharedUnitFields`.

File: app/javascript/hooks/useScript.tsx
Comment (jimlin-sfgov) [reply app/javascript/hooks/useScript.tsx:39]: > What if we just applied eslint rules to the file as needed?

Sounds good!

> this file could use a thorough refactor

Might not be worth the time. This file is only used for our current machine translation process, which is on track to be replaced.

File: app/javascript/api/types/rails/listings/RailsUnit.d.ts
Comment (chadbrokaw) [reply app/javascript/api/types/rails/listings/RailsUnit.d.ts:24]: Okay great, thank you!

File: app/javascript/hooks/useScript.tsx
Comment (chadbrokaw) [reply app/javascript/hooks/useScript.tsx:39]: > Might not be worth the time. This file is only used for our current machine translation process, which is on track to be replaced.

Good point! 

Okay, I will apply the file wide eslint rules. Thanks for the suggestion!

Comment (jimlin-sfgov) [reply app/javascript/hooks/useScript.tsx:39]: fyi this eslint comment can be removed, I see that you removed the other ones already.

**Review by jimlin-sfgov — APPROVED** [review]
great job! this looks good to merge when once we merge https://github.com/SFDigitalServices/sf-dahlia-web/pull/2073

**Pull request merged**
